projects/self_feeding/self_feeding_agent.py

```diff
@@ -302,65 +302,83 @@ def act(self):
+                    self.status = FEEDBACK_REQUESTED
 
                 else:
                     action, reply = self.make_rating_response(rating)
                     self.status = RATING_ACCEPTED
                     # Store just the non-template part of the response in the history
-                    self.history.append(reply)
+                    self.history.update_history({"text": reply})
 
-            elif self.status == EXPLANATION_REQUESTED:
+            elif self.status == FEEDBACK_REQUESTED:
                 action = self.make_action(THANKS + ' ' + NEWTOPIC)
                 self.status = NEWTOPIC_REQUESTED
-                self.history.clear()
+                self.history.reset()
 
             else:
                 raise Exception(f"Unrecognized status: {self.status}")
 
         if self.status == NORMAL:
             action = super().act()
-            self.history.append(action['text'])
+            self.history.update_history(action)
 
         return action
 
     def observe(self, observation):
         """Add to history, concatenate history-size utterances, and add person tokens"""
+
         # If their response is a response to a rating request, no work required
         if self.status == RATING_REQUESTED:
             self.last_rating = observation['text']
         else:
             if 'text' in observation:
-                self.history.append(observation['text'])
+                self.history.update_history(observation)
 
-        if len(self.history) > 0:
+        if len(self.history.history_strings) > 0:
+            # WARNING: Be suspicious of this; is history_size being applied correctly?
             observation['text'] = add_person_tokens(
-                self.history[-self.opt['history_size'] :], last_speaker=1
+                self.history.history_strings[-self.opt['history_size'] :],
+                last_speaker=1,
             )
 
         self.observation = observation
 
         if observation.get('episode_done', True):
-            self.history.clear()
+            self.history.reset()
 
-        return self.vectorize(self.observation, truncate=self.truncate)
+        return self.vectorize(self.observation, self.history)
 
     def batchify(self, observations):
         batch = super().batchify(observations)
         if self.multitask and batch.text_vec is not None:
             # Dialog is the default task; e.g., in interactive mode
             subtasks = [o.get('subtask', 'dialog') for o in observations]
             # Catch error here for debugging
-            if len(set(subtasks)) > 1:
-                import pdb
-
-                pdb.set_trace()
+            assert len(set(subtasks)) == 1
             self.subtask = subtasks[0]
         # print(f"[ context ]: {observations[0]['text']}")
         return batch
 
+    def _set_text_vec(self, obs, history, truncate):
+        """
+        Set the 'text_vec' field in the observation.
+        """
+        if 'text' not in obs:
```

Comment:
@bhancock8 this looks like a near verbatim copy of the TorchAgent `_set_text_vec`. Why is this necessary now?